### PR TITLE
store firmware version info into debugfs for logger usage

### PR DESCRIPTION
**Pull request by xiulipan**
As discussed in https://github.com/thesofproject/soft/issues/111
Stored fw version info into a debugfs to be used with logger
https://github.com/thesofproject/sof/pull/512

### CODE REVIEW

File: sound/soc/sof/loader.c [hunk 1]
```diff
@@ -284,6 +284,12 @@ int snd_sof_run_firmware(struct snd_sof_dev *sdev)
 	init_waitqueue_head(&sdev->boot_wait);
 	sdev->boot_complete = false;
 
+	/* create fw_version debugfs to store boot version info */
+	if (sdev->first_boot)
+		snd_sof_debugfs_create_item(sdev, &sdev->fw_version,
+					    sizeof(sdev->fw_version),
+					    "fw_version");
```
Comment (plbossart) [inline sound/soc/sof/loader.c]: this function can fail and the return status is not checked.
Actually in most cases the code does not check if this function is successful so please submit a PR to clean this up. Either we check the status everywhere or we don't have a status.
Thanks!

File: sound/soc/sof/intel/bdw.c [hunk 2]
```diff
@@ -519,6 +519,10 @@ static int bdw_fw_ready(struct snd_sof_dev *sdev, u32 msg_id)
 		 " Firmware info: version %d:%d-%s build %d on %s:%s\n",
 		 v->major, v->minor, v->tag, v->build, v->date, v->time);
 
+	/* only copy the fw_version into debugfs at first boot */
+	if (sdev->first_boot)
```
Comment (plbossart) [inline sound/soc/sof/intel/bdw.c:523]: need to check that this not NULL in all files, this element may not be created successfully.

File: sound/soc/sof/loader.c (hunk 1 above)
Comment (xiulipan) [reply sound/soc/sof/loader.c]: @plbossart 
Will send another PR to clear debugfs functions.

I also find some thing that related to `snd_sof_debugfs_create_item`
The file ops are binded with the pm options, but when the DSP panic or oops, the PM will fail and make debugfs open/read fail.

File: sound/soc/sof/intel/bdw.c (hunk 2 above)
Comment (xiulipan) [reply sound/soc/sof/intel/bdw.c:523]: `bool first_boot;`
first_boot is a bool value and should be init with 0.
So I did not think it needs to be check

File: sound/soc/sof/intel/bdw.c [hunk 3]
```diff
@@ -519,6 +519,10 @@ static int bdw_fw_ready(struct snd_sof_dev *sdev, u32 msg_id)
...
+		memcpy(&sdev->fw_version, v, sizeof(*v));
+
```
Comment (plbossart) [inline sound/soc/sof/intel/bdw.c:525]: sorry, my comment above was on this line: before copying, make sure you have a valid debugfs element.

Comment (xiulipan) [reply sound/soc/sof/intel/bdw.c:525]: oh, this sdev->fw_version is not a debugfs element. It is a element to sof dev to track the firmware version for onetime.
No need to check if the debugfs creation is success, the mapping is done when we want to read or open it.

Comment (plbossart) [reply sound/soc/sof/intel/bdw.c:525]: ok thanks for the clarification. let's merge when you update on top of the updated branch

**Comment by xiulipan** [comment]
@plbossart 
Rebased to top and fixed the 

> checks on the status and error handling

**Pull request merged**